cognite/client/_api_client.py

```diff
@@ -797,6 +802,88 @@ def _update_multiple(
             return resource_cls._load(updated_items[0], cognite_client=self._cognite_client)
         return list_cls._load(updated_items, cognite_client=self._cognite_client)
 
+    def _upsert_multiple(
+        self,
+        items: CogniteResource | Sequence[CogniteResource],
+        list_cls: Type[T_CogniteResourceList],
+        resource_cls: Type[T_CogniteResource],
+        update_cls: Type[CogniteUpdate],
+        input_resource_cls: Optional[Type[CogniteResource]] = None,
+    ) -> T_CogniteResource | T_CogniteResourceList:
+        is_single = isinstance(items, CogniteResource)
+        items = cast(Sequence[CogniteResource], [items] if is_single else items)
+        try:
+            result = self._update_multiple(items, list_cls, resource_cls, update_cls)
+        except CogniteNotFoundError as not_found_error:
+            items_by_external_id = {item.external_id: item for item in items if item.external_id is not None}
+            items_by_id = {item.id: item for item in items if item.id is not None}
+            # Not found must have an external id as they do no existing in CDF.
+            missing_external_ids = {entry["externalId"] for entry in not_found_error.not_found}
+            to_create = [
+                items_by_external_id[external_id]
+                for external_id in not_found_error.failed
+                if external_id in missing_external_ids
+            ]
+
+            # Updates can have either external id or id. If they have an id, they must exist in CDF.
+            to_update = [
+                items_by_external_id[identifier] if isinstance(identifier, str) else items_by_id[identifier]
+                for identifier in not_found_error.failed
```

Comment:
This list lookup (O(N^2)) should probably be rewritten to linear time complexity